opennlp-tools/src/main/java/opennlp/tools/langdetect/ProbingLanguageDetectorME.java

```diff
@@ -0,0 +1,412 @@
+ * the language detector will stop and report the results.
+ * </p>
+ * <p>
+ * The authors wish to thank Ken Krugler and
+ * <a href="https://github.com/kkrugler/yalder">Yalder</a>}
+ * for the inspiration for many of the design
+ * components of this detector.
+ * </p>
+ *
+ */
+public class ProbingLanguageDetectorME extends LanguageDetectorME {
+
+  /**
+   * Default chunk size (in codepoints) to take from the
+   * initial String
+   */
+  public static final int DEFAULT_CHUNK_SIZE = 300;
+
+  /**
+   * Default minimum consecutive improvements in confidence.
+   * If the best language is the same over this many consecutive
+   * probes, and if the confidence did not go down over those probes,
+   * the detector stops early.
+   */
+  public static final int DEFAULT_MIN_CONSEC_IMPROVEMENTS = 2;
+
+  /**
+   * Default minimum difference in confidence between the language with
+   * the highest confidence and the language with the second highest confidence.
+   */
+  public static final double DEFAULT_MIN_DIFF = 0.10;
+
+  /**
+   * Default absolute maximum length of the String (in codepoints) to process
+   */
+  public static final int DEFAULT_MAX_LENGTH = 10000;
+
+  private static final String SPACE = " ";
+
+  //size at which to break strings for detection (in codepoints)
+  private int chunkSize = DEFAULT_CHUNK_SIZE;
+
+  //require that the "best" language be the same
+  //and that the confidence in that language increase over
+  //this number of probes.
+  private int minConsecImprovements = DEFAULT_MIN_CONSEC_IMPROVEMENTS;
+
+  //Minimum difference in confidence between the best candidate
+  //and the second best candidate
+  private double minDiff = DEFAULT_MIN_DIFF;
+
+  /**
+   * Absolute maximum length (in codepoints) that will processed
+   */
+  private int maxLength = DEFAULT_MAX_LENGTH;
+
+  private CharSequenceNormalizer normalizer;
+
+  /**
+   * Initializes the current instance with a language detector model. Default feature
+   * generation is used.
+   *
+   * @param model the language detector model
+   */
+  public ProbingLanguageDetectorME(LanguageDetectorModel model) {
+    super(model);
+    CharSequenceNormalizer[] normalizers = new CharSequenceNormalizer[] {
+      new EmojiCharSequenceNormalizer(),
+      new UrlCharSequenceNormalizer(),
+      new TwitterCharSequenceNormalizer(),
+      new NumberCharSequenceNormalizer(),
+      new ShrinkCharSequenceNormalizer()
+    };
+
+    this.normalizer = new AggregateCharSequenceNormalizer(normalizers);
+  }
+
+  @Override
+  public Language[] predictLanguages(CharSequence content) {
+    //list of the languages that received the highest
+    //confidence over the last n chunk detections
+    LinkedList<Language[]> predictions = new LinkedList();
+    int start = 0;//where to start the next chunk in codepoints
+    Language[] currPredictions = null;
+    //cache ngram counts across chunks
+    Map<String, MutableInt> ngramCounts = new HashMap<>();
```

Comment:
Can the existing predict method be called here? It should not be duplicated here, if there is something wrong with the way the ngrams are generated the place where this originally happens should be fixed. The generation of the features is done in the Context Generator, and could also be swapped if there are different needs, even by the user at training time.

Reply:
Sounds good.  What I was trying to get away from was the requirement that the ContextGenerator process the full chunk and return String[].  It would be better...at some point... to have the ContextGenerator implement Iterator<String> to enable processing as little as possible.

I think in favor of this incremental, non-breaking change, I can leave the ContextGenerator as is and use the existing predict method.

Reply:
Also not using that ContextGenerator is an option you can implement your own, and then use it instead for training and inference. I see that th einterface is an issue, but this goes down to the model, and other approaches like using hash features would probably the better way to speed it up.

Reply:
Y, agreed on hashing.  

From a probing perspective, moving to `Iterator<String>` would be better than getting all the ngrams from a chunk returned as String[].  One option would be to subclass ContextGenerator with an interface that supports `Iterator<String>` or we could simply add a check in the prober to see if the ContextGenerator happens to support `Iterator<String>`.